File: 프로그래머스/lv3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> board, vector<vector<int>> skill) {
    int answer = 0;
    int N = board.size(), M = board[0].size();
    vector<vector<int>> accum(N + 1, vector<int> (M + 1, 0));
    for(vector<int> s: skill) {
        int type=s[0], r1=s[1], c1=s[2], r2=s[3], c2=s[4], d=s[5];
        if(type == 1) {
            accum[r1][c1] -= d;
            accum[r1][c2 + 1] += d;
            accum[r2 + 1][c1] += d;
            accum[r2 + 1][c2 + 1] -= d;
        }
        else{
            accum[r1][c1] += d;
            accum[r1][c2 + 1] -= d;
            accum[r2 + 1][c1] -= d;
            accum[r2 + 1][c2 + 1] += d;
        }
    }
    
    for(int i = 0; i < N; i++) {
        for(int j = 1; j < M; j++) {
            accum[i][j] += accum[i][j - 1];
        }
    }
    
    for(int j = 0; j < M; j++) {
        for(int i = 1; i < N; i++) {
            accum[i][j] += accum[i - 1][j];
        }
    }
    
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < M; j++) {
            board[i][j] += accum[i][j];
            if(board[i][j] > 0) answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.py
def solution(board, skill):
    # 카카오 해설 봄 -> 2차원 배열 누적합
    N = len(board)
    M = len(board[0])
    
    accum = [[0] * (M+1) for _ in range(N+1)]
    
    for cur_skill in skill:
        skill_type, r1, c1, r2, c2, degree = cur_skill
        
        if skill_type == 1:
            degree = -degree
        
        accum[r1][c1] += degree
        accum[r1][c2 + 1] -= degree
        accum[r2 + 1][c1] -= degree
        accum[r2 + 1][c2 + 1] += degree
    
    # 행별 누적합 계산
    for i in range(N):
        for j in range(1, M):
            accum[i][j] += accum[i][j - 1]
    
    # 열별 누적합 계산
    for j in range(M):
        for i in range(1, N):
            accum[i][j] += accum[i - 1][j]
    
    answer = 0
    for i in range(N):
        for j in range(M):
            board[i][j] += accum[i][j]
            answer = answer + 1 if board[i][j] > 0 else answer
        
    return answer